site/app/models/gradeable/Gradeable.php

```diff
@@ -0,0 +1,446 @@
+        }
+
+        $this->ta_view_start_date = $dates['ta_view_start_date'];
+        $this->grade_start_date = $dates['grade_start_date'];
+        $this->grade_released_date = $dates['grade_released_date'];
+        $this->grade_locked_date = $dates['grade_locked_date'];
+
+        if($this->type === GradeableType::ELECTRONIC_FILE) {
+            if($this->team_assignment) {
+                $this->team_lock_date = $dates['team_lock_date'];
+            }
+            $this->submission_open_date = $dates['submission_open_date'];
+            $this->submission_due_date = $dates['submission_due_date'];
+            $this->late_days = $dates['late_days'];
+        }
+    }
+    private function setTaViewStartDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setGradeStartDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setGradeReleasedDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setGradeLockedDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setTeamLockDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setSubmissionOpenDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+    private function setSubmissionDueDate($date)
+    {
+        throw new NotImplementedException('Individual date setters are disabled, use "setDates" instead');
+    }
+
+    private function setId($id)
+    {
+        preg_match('/^[a-zA-Z0-9_-]*$/', $id, $matches, PREG_OFFSET_CAPTURE);
+        if (count($matches) === 0) {
+            throw new \InvalidArgumentException("Gradeable id must be alpha-numeric/hyphen/underscore only");
+        }
+        $this->id = $id;
+    }
+
+    public function setTitle($title)
+    {
+        if ($title === '') {
+            throw new \InvalidArgumentException("Gradeable title must not be blank");
+        }
+        $this->title = strval($title);
+    }
+
+    private function setType($type)
+    {
+        // Call this to make an exception if the type is invalid
+        GradeableType::typeToString($type);
+        $this->type = $type;
+    }
+
+    public function setMinGradingGroup($group)
+    {
+        // Disallow the 0 group (this may catch some potential bugs with instructors not being able to edit gradeables)
+        if (is_int($group) && $group > 0) {
+            $this->min_grading_group = $group;
+        } else {
+            throw new \InvalidArgumentException("Grading group must be a natural number");
```

Comment:
> natural number

I know some mathematicians who would argue with this... Just say must be an integer greater than zero